### PR TITLE
[Linx Impulse]: Add Linx Impulse integration

### PR DESCRIPTION
**Pull request by vitoUwu**

### CODE REVIEW

**Comment by vitoUwu** [comment]
https://github.com/deco-cx/apps/pull/397#discussion_r1494678444

Fixed

**Comment by vitoUwu** [comment]
https://github.com/deco-cx/apps/pull/397#discussion_r1494680113 @mcandeia 

During testing, I noticed that I was getting some errors in the terminal, when I realized that some requests with "_frsh" were causing these errors, I'm not sure what could have caused this.

**Comment by guitavano** [comment]
Very nice!

As this is a full integration, it would be nice if you can put here two things:

- a demo store working (you probably did one to create)
- a loom explaining in general how it was made and things that we should know

**Comment by vitoUwu** [comment]
> Very nice!
> 
> As this is a full integration, it would be nice if you can put here two things:
> 
> * a demo store working (you probably did one to create)
> * a loom explaining in general how it was made and things that we should know

For example, I have https://deco-sites-clovis.deno.dev/ , we use product shelves with the Linx Impulse loader, we use autocomplete in the search bar, and we use click and page view event tracking


https://github.com/deco-cx/apps/assets/32278696/a35f95c8-d9a6-4d78-a105-75220473a8a8

File: linx-impulse/mod.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+   */
+  secretKey?: string;
+  /**
+   * @title Origin URL
+   * @description Set the origin URL to authenticate the request if not using the secret key
+   */
+  origin?: string;
+  /**
+   * @title Image CDN URL
+   * @description e.g.: https://{account}.myvtex.com/
```
Comment (guitavano) [inline linx-impulse/mod.ts:34]: Is myvtex url a good example? Will people use that?

File: linx-impulse/mod.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    headers: new Headers({ "Accept": "application/json" }),
+  });
+
+  const chaordicApi = createHttpClient<ChaordicAPI>({
+    base: "https://recs.chaordicsystems.com/",
+    headers: new Headers({ "Accept": "application/json" }),
+  });
+
+  const state = {
+    ...props,
```
Comment (guitavano) [inline linx-impulse/mod.ts:67]: Let`s choose which props should be in state? I really appreciate use secretKey and appKey only to create the clients.

File: linx-impulse/loaders/products/productList.ts [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+/**
+ * @title Category Page
+ */
+interface CategoryPageProps extends BaseProps {
+  /**
+   * @hide
+   */
+  page: "category";
+  /**
+   * @description It is recommended to create a global loader so that the loader is not called more than once
```
Comment (guitavano) [inline linx-impulse/loaders/products/productList.ts:69]: Nice!

File: linx-impulse/loaders/products/productList.ts [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+    .replace(/[\u0300-\u036f]/g, "");
+
+/**
+ * @title Page
+ */
+type Props =
+  | HomePageProps
+  | ProductPageProps
+  | CategoryPageProps
+  | SearchPageProps;
```
Comment (guitavano) [inline linx-impulse/loaders/products/productList.ts:110]: Bad say that, but we already have problems doing that, can you change to:
```suggestion
type Props = { props: HomePageProps | ProductPageProps | CategoryPageProps | SearchPageProps;
```

**Comment by guitavano** [comment]
Only to confirm, linx-impulse app will always be used with another main ecommerce integration, right? It's why we do not need to create checkout proxy, for example?

File: linx-impulse/mod.ts [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+/**
+ * @title Linx Impulse
+ */
+export default function Linx(props: State) {
+  const eventsApi = createHttpClient<EventsAPI>({
+    base: "https://api.event.linximpulse.net/",
+    headers: new Headers({
+      "Accept": "application/json",
+      "User-Agent":
+        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36",
```
[the remark this answers is not in the document]
Comment (vitoUwu) [reply linx-impulse/mod.ts]: Forgot to remove, fixed

File: linx-impulse/mod.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+/** @title LINX Impulse */
+export interface State {
+  /**
+   * @title Api Key
+   */
+  apiKey: string;
+  /**
+   * @title Secret Key
+   */
+  secretKey?: string;
```
[the remark this answers is not in the document]
Comment (vitoUwu) [reply linx-impulse/mod.ts]: Fixed!

File: linx-impulse/mod.ts (hunk 1 above)
Comment (vitoUwu) [reply linx-impulse/mod.ts:35]: It's the same example used in the Linx integration, stores that use linx + vtex have the cdn url like this

File: linx-impulse/mod.ts (hunk 2 above)
Comment (vitoUwu) [reply linx-impulse/mod.ts:65]: According to the Linx Impulse documentation, the apiKey and secretKey are sent in the body in POST requests, and in the query params in GET requests, can I define this when creating the client? The apiKey is a little more public, when integrated via script it is sent to the client side

**Comment by vitoUwu** [comment]
> Only to confirm, linx-impulse app will always be used with another main ecommerce integration, right? It's why we do not need to create checkout proxy, for example?

@guitavano Yes, it is mandatory to have a main integration, be it Linx Commerce, VTEX, etc...

File: linx-impulse/mod.ts (hunk 2 above)
Comment (guitavano) [reply linx-impulse/mod.ts:65]: It is possible to add on headers, I checked and should work:

image

**Review by guitavano — APPROVED** [review]

**Pull request merged**